x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/cache/ProgressListenableActionFuture.java

```diff
@@ -0,0 +1,203 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+package org.elasticsearch.index.store.cache;
+
+import org.elasticsearch.action.ActionFuture;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.AdapterActionFuture;
+import org.elasticsearch.common.collect.Tuple;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+/**
+ * An {@link ActionFuture} that listeners can be attached to. Listeners are executed when the future is completed
+ * or when a given progress is reached. Progression is updated using the {@link #onProgress(long)} method.
+ *
+ * Listeners are executed within the thread that triggers the completion, the failure or the progress update and
+ * the progress value passed to the listeners on execution is the last updated value.
+ */
+class ProgressListenableActionFuture extends AdapterActionFuture<Long, Long> {
+
+    protected final long start;
+    protected final long end;
+
+    // modified under 'this' mutex
+    private volatile List<Tuple<Long, ActionListener<Long>>> listeners;
+    protected volatile long progress;
+    private volatile boolean completed;
+
+    /**
+     * Creates a {@link ProgressListenableActionFuture} that accepts the progression
+     * to be within {@code start} (inclusive) and {@code end} (exclusive) values.
+     *
+     * @param start the start (inclusive)
+     * @param end   the end (exclusive)
+     */
+    ProgressListenableActionFuture(long start, long end) {
+        super();
+        this.start = start;
+        this.end = end;
+        this.progress = start;
+        this.completed = false;
+        assert invariant();
+    }
+
+    private boolean invariant() {
+        assert start < end : start + " < " + end;
+        synchronized (this) {
+            assert completed == false || listeners == null;
+            assert start <= progress : start + " <= " + progress;
+            assert progress <= end : progress + " <= " + end;
+            assert listeners == null || listeners.stream().allMatch(listener -> progress < listener.v1());
```

Comment:
Should we require `completed == false || progress == end` too?

Reply:
I don't think so: `completed` indicates that the future is done, either with success or failure. In case of failure it could be completed before the `progress` reached `end`.

Reply:
Ah yes so it does. Can we assert that successful completion only happens with `progress == end`?

Reply:
> Can we assert that successful completion only happens with `progress == end`?

It means to change the ` done()`  method to pass around the completion state (success or failure/cancel) of the future so that we can later compare the successful completion plus the `progress`/`end` values. It also means that the progress must be updated in a more strictly manner before completing the future. I gave it a try in a5b29d5c1233e39e1d870f6dd58d6b844964772a, let me know what you think.